code/xcvm/cosmwasm/contracts/gateway/src/common.rs

```diff
@@ -1,21 +1,58 @@
 use crate::{
-	error::ContractError,
-	state::{Config, CONFIG, ROUTER},
+	error::{ContractError, ContractResult},
+	state,
 };
-use cosmwasm_std::Deps;
 
-pub fn ensure_admin(deps: Deps, sender: &str) -> Result<(), ContractError> {
-	let Config { admin, .. } = CONFIG.load(deps.storage)?;
-	if &admin == sender {
+use cosmwasm_std::{Deps, Env, Event, MessageInfo};
+use xc_core::InterpreterOrigin;
+
+/// Creates an event with contract’s default prefix and given action attribute.
+pub(crate) fn make_event(action: &str) -> Event {
+	Event::new(cw_xc_common::gateway::EVENT_PREFIX).add_attribute("action", action)
+}
+
+/// Ensure that the sender of the message is current contract.
+///
+/// Address of the current contract is read from the environment passed in `env`
+/// and sender of the message is read from message info passed in `info`
+/// argument.
+pub(crate) fn ensure_self(env: &Env, info: &MessageInfo) -> ContractResult<()> {
+	if info.sender == env.contract.address {
+		Ok(())
+	} else {
+		Err(ContractError::NotAuthorized)
+	}
+}
+
+/// Ensure that the sender of the message is an interpreter for the provided
+/// `interpreter_origin`.
+///
+/// Sender of the message is read from message info passed in `info` argument.
+/// Address of the interpreter is read from storage (see
+/// [`state::INTERPRETERS`]).  Check fails if interpreter is not known or its
+/// address is not known
+pub(crate) fn ensure_interpreter(
```

Comment:
Why not having one function that deals with all type of authorizations, and an enum for the different types?

Reply:
Each call site would always call it with the same authorisation type so this doesn’t feel like it makes things better.